### ISSUE TITLE
Stack overflow when checking recursive constraint

### ISSUE THREAD

**Issue by yallop** [ocaml/ocaml#9866]
The following program

```ocaml
type 'a t = 'b  constraint 'a = 'b t
```

is correctly rejected with an error message in 4.11.0

```
Line 1, characters 0-36:
1 | type 'a t = 'b  constraint 'a = 'b t;;
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: In this definition, a type variable cannot be deduced
       from the type parameters.
```

but causes a stack overflow in the compiler in `trunk`.

@garrigue: I think this was introduced by #9500.

**Comment by garrigue** [comment]
This looks like another version of #9314 .
The error message in 4.11 seems only remotely related to what happens.

The same stack overflow occurs also in 4.00.
In 3.12.2, the error is even stranger:
```
# type 'a t = 'b  constraint 'a = 'b t;;
# let f x = (x : 'a t :> int);;
Fatal error: exception Assert_failure("typing/ctype.ml", 1149, 6)
```

I'm not sure what to do about this one.
Should we just reject constraints that result in extracting the body from a parameter?
This might also solve #9314, which would be great.

**Comment by gasche** [comment]
What do you mean by "the body" in "extracting the body from a parameter"?

Note: the example `type 'a gamma = 'b constraint 'a = < gamma : 'b >` from #9314 is a projection operator for a type-level record; I'm worried that disallowing this would make type-level records impractical in OCaml, and I've written libraries that use type-level records in ways that seem useful -- removing them would reduce expressivenes in ways that are probably still used in the wild.

**Comment by garrigue** [comment]
Actually I found a solution (#9870) that doesn't introduce any restriction.
Interestingly, the problem here is that 4.12 is able to detect that `t` is injective, but unification could not cope with this definition. This is fixed with only minor side-effects, but as everything related to constrained parameters, it is hard to define what is correct. Again, there would be no problem if all types were always fully expanded, so a full solution could be left for the future.

**Comment by yallop** [comment]
> The error message in 4.11 seems only remotely related to what happens.

Here's a similar case where the connection between the message and the code seems even more tenuous:

```
# type 'a t = 'b constraint 'a = ('b t * 'b) ;;
Line 1:
Error: Type declarations do not match:
         type 'a t = 'b constraint 'a = 'b t * 'b
       is not included in
         type 'a t = 'b constraint 'a = 'b t * 'b
```

**Comment by garrigue** [comment]
With #9870, this gives:
```
# type 'a t = 'b constraint 'a = ('b t * 'b) ;;
Error: Constraints are not satisfied in this type.
       Type 'b t should be an instance of ('a t * 'a) t
```
This is a bit more meaningful.

**Comment by garrigue** [comment]
Unfortunately, #9870 does not fix the problem for recursive modules. @yallop found the following counter-example.

```ocaml
module type S =
sig
  type !'a s
  type !'a t = 'b  constraint 'a = 'b s
end

module rec M : S =
struct
  type !'a s = 'a M.t
  type !'a t = 'b constraint 'a = 'b s
end
```

**Comment by gasche** [comment]
I can reproduce the bug on trunk. Reopening.

@garrigue would you maybe consider storing this known-issue in the testsuite? (Stack-overflow tests on native are fragile, and will change with Multicore; I would write a test that only calls the bytecode compiler, with the proper stack-size flag to ensure that it reliably stack overflows.)

**Comment by gasche** [comment]
The example proposed by @yallop above is rejected as it should in 5.0 and trunk. However I believe that the underlying issue still exists. At least, this was my understanding in April 2021 in https://github.com/ocaml/ocaml/pull/9324#issuecomment-822187379; quoting myself:

> I am not sure what the current status is for this class of issues. My impression from testing trunk is that the simple examples reported in #9866 were fixed by #9870, but the complex examples involving recursive modules and also the examples in #9314 are still failing.

If I understand correctly, #11648 is supposed to help / fix the root cause, but reviewing it looks like a mouthful and I personally haven't had the time yet.

I propose to close the present issue, whose example was fixed, as superseded by #9314 which is also an open issue whose examples are still failing on trunk.
